enhancements/installer/single-node-installation-bootstrap-in-place.md

```diff
@@ -0,0 +1,483 @@
+
+We could have the installer generate an Ignition config that includes
+all of the assets required for launching the single node cluster
+(including TLS certificates and keys).  When booting a machine with
+CoreOS and this Ignition configuration, the Ignition config would lay
+down the control plane operator static pods and create a static pod
+that functions as `cluster-bootstrap` This pod should delete itself
+after it is done applying the OCP assets to the control plane.
+
+### Preserve etcd database instead of a snapshot
+
+Another option for preserving the etcd database when pivoting from
+bootstrap to production is to copy the entire database, instead of
+using a snapshot operation.  When stopped, etcd will save its state
+and exit. We can then add the `/var/lib/etcd` directory to the master
+Ignition config.  After the reboot, etcd should start with all the
+data it had prior to the reboot. By using a snapshot, instead of
+saving the entire database, we will have more flexibility to change
+the production etcd configuration before restoring the content of the
+database.
+
+### Creating a bootable installation artifact directly from the installer
+
+In order to embed the bootstrap-in-place-for-live-iso Ignition config
+to the liveCD the user needs to download the liveCD image and the
+`coreos-installer` binary.  We considered adding an `openshift-install
+create single-node-iso` command that that result a liveCD with the
+`bootstrap-in-place-for-live-iso.ign` embeded.  The installer command
+could also include custom manifests, especially `MachineConfig`
+instances for setting the realtime kernel, setting kernel args,
+injecting network configuration as files, and choosing the target disk
+drive for `coreos-installer`.  Internally, `create single-node-iso`
+would compile a single-node-iso-target.yaml into Ignition (much like
+coreos/fcct) and include it along with the Ignition it generates and
+embed it into the ISO.
+
+This approach has not been rejected entirely, and may be handled with
+a future enhancement.
+
+### Allow bootstrap-in-place in cloud environment (future work)
+For bootstrap-in-place model to work in cloud environemnt we need to mitigate the following gaps:
+1. The bootstrap-in-place model relay on the live ISO environment as a place to write bootstrapping files so that they don't end up on the real node.
+Optional mitigation: We can mimic this environment by mounting some directories as tmpfs during the bootstrap phase.
+2. The bootstrap-in-place model uses coreos-installer to write the final Ignition to disk along with the RHCOS image.
+Optional mitigation: We can boot the machine with the right RHCOS image for the release.
+Instead of writing the Ignition to disk we will use the cloud credentials to update the node Ignition config in the cloud provider.
+
+### Use `create ignition-configs` with environment variable to generate the `bootstrap-in-place-for-live-iso.ign`.
```

Comment:
This section describes 2 alternatives. See https://github.com/eranco74/enhancements/pull/5